### DISCUSSION TITLE
Get database name inside migrator script

### DISCUSSION THREAD

**Discussion by fairking** [fluentmigrator/fluentmigrator#1522] [Q&A]
I am planning to set [read committed snapshot isolation level](https://stackoverflow.com/a/1664236/1143349) on mssql db via migrator by adding the following line:
``` c#
IfDatabase("SqlServer").Execute.Sql("ALTER DATABASE ???YourDatabaseName??? SET READ_COMMITTED_SNAPSHOT ON;");
```

The problem is I don't have a database name. I can only parse the connection string (see example bellow):
``` c#
[Migration(20210928_1000, transactionBehavior: TransactionBehavior.None)]
public class MIG_20210928_1000_ReadCommitted_Snapshot_Mssql : Migration
{
    public override void Up()
    {
        var dbName = ConnectionString.GetDbName();
        IfDatabase("SqlServer").Execute.Sql($"ALTER DATABASE [{dbName}] SET SINGLE_USER WITH ROLLBACK IMMEDIATE;");
        IfDatabase("SqlServer").Execute.Sql($"ALTER DATABASE [{dbName}] SET READ_COMMITTED_SNAPSHOT ON;");
        IfDatabase("SqlServer").Execute.Sql($"ALTER DATABASE [{dbName}] SET MULTI_USER;");
    }

    public override void Down()
    {
        throw new NotImplementedException();
    }
}
```
``` c#
public static string GetDbName(this string connectionString)
{
    var connection = new SqlConnection(connectionString);
    return connection.Database;
}
```

In NHibernate having a session I can easlily use the following:
``` c#
_session.Connection.Database
```
But inside the migrator context I don't have a NH session nor database name. 

Is there a way to get the database name besides manually parsing the connection string?

**Comment by mburnell** [comment]
I found myself with a similar requirement. I figured there'd be an active connection / transaction, and I could retrieve the database name from the environment to work around the issue using something like this:

```
declare @database nvarchar(255) = (select DB_Name())
EXEC('ALTER DATABASE ' + @database + ' READ_COMMITTED_SNAPSHOT ON')
```

Unfortunately, this doesn't work, as alter database statements aren't allowed in multi-statement transactions. Given the transactionality of migrations is one of their great strengths, I ended up just accepting that this logic was probably a poor fit for a migration.

**Reply by jzabroski** [reply]
I responded below.

**Comment by jzabroski** [comment]
I'm sorry I never responded to this.

I use Execute.WithConnection for these things.  You just have to bare in mind that Execute.WithConnection flows differently than other migrations since it is a promise to execute something. But you can basically stack Execute.WithConnection commands and thread them together. I create a separate layer for this stuff so it's not directly in my migrations and is a re-usable library that just uses a SqlConnection or DbConnection, so that I can re-use this stuff in many different projects.
